Reland "RegisterCoalescer: Add implicit-def of super register when coalescing SUBREG_TO_REG"

sdesmalen-arm opened the pull request:
A SUBREG_TO_REG instruction expresses that the top bits of the result register are set to a certain value (e.g. 0).

The example below expresses that the result of %1 will have the top 32 bits zeroed and the lower 32bits being equal to the result of INSTR.
```
    %0:gpr32 = INSTR
    %1:gpr64 = SUBREG_TO_REG 0, %0, sub32
```
When the RegisterCoalescer tries to remove SUBREG_TO_REG instructions by coalescing %0 into %1, it must keep the same semantics. Currently however, the RegisterCoalescer would emit:
```
    %1.sub32:gpr64 = INSTR
```
which no longer expresses that the top 32-bits of the register are defined (zeroed) by INSTR.

This may cause issues with e.g. machine copy propagation where the pass may think it can remove a COPY-like instruction because the MIR says only the bottom 32-bits are defined/used, even though other uses of the register rely on the top 32-bits being zeroed by the COPY-like instruction.

This PR changes the RegisterCoalescer to instead emit:
```
    undef %1.sub32:gpr64 = MOVimm32 42, implicit-def %1
```
to express that the entire contents of %1:gpr64 are defined by the instruction.

This tries to reland #134408 which had to be reverted to a few reported failures.

sdesmalen-arm committed:
Revert "Revert "Reland "RegisterCoalescer: Add implicit-def of super register when coalescing SUBREG_TO_REG" (#134408)""

This reverts commit ed5bd23867eacaa3789060f9250ba6fcece2a3de.

sdesmalen-arm committed:
Fixes for issues reported when landing #134408

The register coalescer ran into some asserts:

* The newly added code tried to get the LiveInterval for a physical
  register (unguarded path)

* The assert 'assert(SubregToRegSrcInsts.empty() && "can this happen?");'
  could happen when using SUBREG_TO_REG to say that the top bits
  of the second register in a {128, 128} register tuple are zero, e.g.

  %8.qsub1:qq = MOVIv2d_ns 0
  %4:zpr = SUBREG_TO_REG 0, %8.qsub1:qq, %subreg.zsub

sdesmalen-arm committed:
Fix an issue where we only add an implicit-def if there is a sub-idx

There's no point in doing it for something like this:

    %1:gpr64sp = ORRXri %0, 8128
    %3:gpr64 = SUBREG_TO_REG 0, %1.sub_32, %subreg.sub_32

where ORRXri already writes to a gpr64 register class (although one that
also includes the sp).

sdesmalen-arm committed:
addDeadDef should iterate over all defs (including implicit ones)

otherwise it results in a machine verifier failure when
splitting the liverange and rematerialising a MOVi32imm
instruction.

This is only an issue when enabling subreg liveness tracking.

MacDue commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
```diff
@@ -1935,6 +1954,79 @@ void RegisterCoalescer::updateRegDefsUses(Register SrcReg, Register DstReg,
+        //  48B    %0:gpr64 = SUBREG_TO_REG 0, %1, sub32
+        //
+        //  Only the MOVIMM32 require a def of the top lanes and any intervals
+        //  for the top 32-bits of the def at 16B should be removed.
```
```suggestion
        //  Only the MOVIMM32 requires a def of the top lanes and any intervals
        //  for the top 32-bits of the def at 16B should be removed.
```

MacDue commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
```diff
@@ -2046,6 +2138,30 @@ void RegisterCoalescer::setUndefOnPrunedSubRegUses(LiveInterval &LI,
 
+/// For a given use of value \p Idx, it returns the def in the current block,
+/// or otherwise all possible defs in preceding blocks.
+static bool FindDefInBlock(SmallPtrSetImpl<MachineBasicBlock *> &VisitedBlocks,
```
```suggestion
static bool findDefInBlock(SmallPtrSetImpl<MachineBasicBlock *> &VisitedBlocks,
```

MacDue reviewed (commented):
I think the new fixes look good :+1: 

I just have some general questions/nits about the initial patch (as I didn't review it before, and it's a fairly tricky change). If @arsenm is happy with the change as-is, you can skip my suggestions. 

For the PR description, I think it'd be good to include the initial description, since that'd help explain the change in a git blame. Also, not sure if the commit references should be included, since they are not commits in the LLVM repo. Maybe just explain the changes from the initial PR directly in the description.

sdesmalen-arm commented on llvm/lib/CodeGen/RegisterCoalescer.cpp:
```diff
@@ -1935,6 +1954,79 @@ void RegisterCoalescer::updateRegDefsUses(Register SrcReg, Register DstReg,
+          assert((SR.LaneMask & UnusedLanes) == SR.LaneMask &&
+                 "Unexpected lanemask. Subrange needs finer granularity");
+
+          SlotIndex UseIdx = LIS->getInstructionIndex(*UseMI).getRegSlot(false);
```
done.

jayfoad commented:
> Try to reland #134408, which had to be reverted to a few reported failures.
> 
> The patches [7228ceb](https://github.com/llvm/llvm-project/commit/7228ceb9749d9491710bfa3f872c261942ff15bc) and [9478968](https://github.com/llvm/llvm-project/commit/9478968f2262136f5fe3a738101885de0daef6c7) each fix issues that were reported (with corresponding reproducers/tests). [aebc086](https://github.com/llvm/llvm-project/commit/aebc086b73d15232bfd0c8e8f8099343b28c91a4) was found when enabling subreg liveness tracking for AArch64 (still disabled by default).
> 
> Given the small changes to the original PR, I wasn't sure whether I could just push it, but I figured I'd play it safe instead.

Just a high level comment: since this is now the 8th attempt to land this patch over more than 2 years, would it make sense to restate what this patch does and why? Otherwise you have to follow a very long trail of breadcrumbs to find any kind of explanation.

sdesmalen-arm committed:
Address suggestions

sdesmalen-arm committed:
Apply suggestion, and fix test failures

It should have started the worklist with the predecessors of MBB,
not MBB itself.

MacDue reviewed (approved):
The new fixes make sense to me :+1:  
Just one more question about the initial patch:

sdesmalen-arm committed:
Use SmallPtrSet instead of SmallVector/ArrayRef

sdesmalen-arm committed:
Avoid adding duplicate implicit-defs
